### PR TITLE
[Merged by Bors] - feat(NNRat): More basic lemmas

### PR DESCRIPTION
**Pull request by YaelDillies**
From LeanAPAP

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/NNRat/Defs.lean
Comment (b-mehta) [inline Mathlib/Data/NNRat/Defs.lean:59]: Let's name the lemma in question here: which is the specific one that you want to use it earlier than?

**Review by b-mehta — APPROVED** [review]
The changes here appear sensible and I agree they are useful in practice. I'd like someone else to double-check the instance changes are usable.

maintainer merge

**Comment by YaelDillies** [comment]
Probably @eric-wieser wants to have a look

**Comment by mattrobball** [comment]
I am not a fan of the import changes.

File: Mathlib/Data/Rat/Cast/Lemmas.lean
Comment (mattrobball) [inline Mathlib/Data/Rat/Cast/Lemmas.lean:7]: This seems like a big change in the import structure.

File: Mathlib/Data/Rat/Cast/Defs.lean
Comment (mattrobball) [inline Mathlib/Data/Rat/Cast/Defs.lean:11]: Given the current design, I don't really want `NNRat` to go first before `Rat`.

File: Mathlib/Algebra/Algebra/Rat.lean
Comment (eric-wieser) [inline Mathlib/Algebra/Algebra/Rat.lean:51]: I think you need to inline `castHom` here, otherwise you get defeq-but-not-reducible-defeq diamonds. (cc @mattrobball)

File: Mathlib/Algebra/Algebra/Rat.lean
Comment (eric-wieser) [inline Mathlib/Algebra/Algebra/Rat.lean:54]: Arguably, `RingHomClass.toNNRatLinearMapClass` would be a more accurate name.

File: Mathlib/Algebra/Algebra/Rat.lean
Comment (mattrobball) [reply Mathlib/Algebra/Algebra/Rat.lean:51]: I am guessing that `castHom` semi-reducible?

File: Mathlib/Algebra/Algebra/Rat.lean
Comment (YaelDillies) [reply Mathlib/Algebra/Algebra/Rat.lean:54]: Yes, true

File: Mathlib/Algebra/Algebra/Rat.lean
Comment (YaelDillies) [reply Mathlib/Algebra/Algebra/Rat.lean:51]: Oh, because `Algebra` also implements a coercion... I will make `castHom` reducible, since that's how we use it in practice anyway

File: Mathlib/Data/Rat/Cast/Defs.lean
Comment (YaelDillies) [reply Mathlib/Data/Rat/Cast/Defs.lean:11]: Sorry, I don't understand this complaint. `NNRat` should always be earlier than `Rat` since it's used in `Semifield` while `Rat` is used in `Field` only.

Oh I guess you mean "the current design of `NNRat` as a subtype of `Rat`" in which case I shall reply "That design needs to be changed for independent reasons anyway, and I am willing to make it happen soon".

File: Mathlib/Data/NNRat/Defs.lean
Comment (YaelDillies) [reply Mathlib/Data/NNRat/Defs.lean:60]: Uh, maybe you're right, but I have been using this `nolint simpNF` pattern for a while so there will be a lot of occurrences to be fixed if you truly believe this is a better solution.

File: Mathlib/Algebra/Algebra/Rat.lean
Comment (YaelDillies) [reply Mathlib/Algebra/Algebra/Rat.lean:41]: I created a new file under `test` and wrote
```
import Mathlib.Algebra.Algebra.Rat

/-- The two `Algebra ℚ ℚ` instances should coincide. -/
example : algebraRat = Algebra.id ℚ := rfl
```
and it failed 😨 

Turns out `algebraRat` was a bound variable as we haven't turned off auto-implicit in the `test` folder 😅

File: Mathlib/Data/Rat/Cast/Lemmas.lean
Comment (YaelDillies) [reply Mathlib/Data/Rat/Cast/Lemmas.lean:7]: I know why this is necessary and I will say two things:
1. It's avoidable in a way that doesn't involve unnecessarily splitting files
2. I am working on this way, but it's definitely a later PR and I don't really want to get into the weeds of that now

File: Mathlib/Data/Rat/Cast/Defs.lean
Comment (mattrobball) [reply Mathlib/Data/Rat/Cast/Defs.lean:11]: Subtype of `Rat` is fine in and of itself. But `NNRat.Defs` imports way too much bundled ordered algebra classes to do what it needs to do. I'd prefer that gets fixed before we add more on top of it.

Comment (YaelDillies) [reply Mathlib/Data/Rat/Cast/Defs.lean:11]: `Data.NNRat.Defs` is already imported in `Data.Rat.Cast.Defs` on master, so I do not understand your comment

Comment (mattrobball) [reply Mathlib/Data/Rat/Cast/Defs.lean:11]: Sorry, I kinda lost this. I already have a PR disconnecting the basics of `NNRat.Defs` #15072 from bundled ordered algebra that this might conflict with. Can we wait until that gets sorted out?

File: Mathlib/Data/NNRat/Defs.lean
Comment (eric-wieser) [inline Mathlib/Data/NNRat/Defs.lean:40]:
```suggestion
1. There is a significant portion of the library which needs the early lemma to be available via
   `simp` and which doesn't have access to the more general lemmas.
2. The more general lemmas have more complicated typeclass assumptions, causing rewrites with them
   to be slower.
```

File: Mathlib/Data/NNRat/Defs.lean
Comment (eric-wieser) [inline Mathlib/Data/NNRat/Defs.lean:36]:
```suggestion
early lemma to be tagged `@[simp high]` (rather than `@[simp, nolint simpNF]` or un`@[simp]`ed):
```

File: Mathlib/Data/NNRat/Defs.lean
Comment (eric-wieser) [inline Mathlib/Data/NNRat/Defs.lean:33]: I think we usually avoid abbreviations here?
```suggestion
library_note "specialised high priority simp lemma" /--
```
and at the references.

File: Mathlib/Data/NNRat/Defs.lean
Comment (eric-wieser) [inline Mathlib/Data/NNRat/Defs.lean:34]:
```suggestion
It sometimes happens that a simp lemma declared early in the library can be proved by `simp` using
```

File: Mathlib/Data/Rat/Cast/CharZero.lean
Comment (eric-wieser) [inline Mathlib/Data/Rat/Cast/CharZero.lean:47]: On second thoughts, let's leave this can of worms closed for now:
```suggestion
def castHom : ℚ →+* α where
```
We can fix it in a later PR.

File: Mathlib/Data/Rat/Cast/CharZero.lean
Comment (eric-wieser) [inline Mathlib/Data/Rat/Cast/CharZero.lean:95]: idem
```suggestion
def castHom : ℚ≥0 →+* α where
```

File: Mathlib/Data/Rat/Cast/Defs.lean
Comment (eric-wieser) [inline Mathlib/Data/Rat/Cast/Defs.lean:266]:
```suggestion
/-- If monoid with zero homs `f` and `g` from `ℚ` agree on the integers then they are equal.
```

File: Mathlib/Data/Rat/Cast/Defs.lean
Comment (eric-wieser) [inline Mathlib/Data/Rat/Cast/Defs.lean:274]:
```suggestion
/-- If monoid with zero homs `f` and `g` from `ℚ` agree on the positive naturals and `-1` then
```

File: Mathlib/Data/Rat/Cast/Lemmas.lean
Comment (eric-wieser) [inline Mathlib/Data/Rat/Cast/Lemmas.lean:25]: Do we have the `zpow` version somewhere else?

File: test/algebra_rat.lean
Comment (eric-wieser) [inline test/algebra_rat.lean:1]: Can you put this in the instance_diamonds folder?

**Review by eric-wieser — APPROVED** [review]
My preference would be to put this in before #15072; @YaelDillies, are you happy to help resolve conflicts in the other PR if they go in that order?

File: Mathlib/Data/NNRat/Defs.lean
Comment (YaelDillies) [reply Mathlib/Data/NNRat/Defs.lean:40]: Uh, that doesn't respect the "indent by two spaces" rule. We don't right-flush anything else, eg in `calc` syntax

File: Mathlib/Data/Rat/Cast/Lemmas.lean
Comment (YaelDillies) [reply Mathlib/Data/Rat/Cast/Lemmas.lean:25]: Yes, `NNRat.cast_zpow` in `Data.Rat.Cast.CharZero`

**Comment by YaelDillies** [comment]
> My preference would be to put this in before #15072; @YaelDillies, are you happy to help resolve conflicts in the other PR if they go in that order?

That's my preference too, especially since there is so much of APAP waiting on this PR to be merged. I am of course happy to resolve conflicts in the other PR

File: Mathlib/Data/NNRat/Defs.lean
Comment (eric-wieser) [reply Mathlib/Data/NNRat/Defs.lean:41]: Ultimately what matters is whether our markdown formatter allows this.

... which I guess also doesn't matter because `library_note`s don't yet render anywhere:.

File: Mathlib/Data/Rat/Cast/CharZero.lean
Comment (eric-wieser) [inline Mathlib/Data/Rat/Cast/CharZero.lean:84]: I think this might be `eq_iff'`?

**Review by eric-wieser — APPROVED** [review]
bors d+

File: Mathlib/Data/Rat/Cast/CharZero.lean
Comment (YaelDillies) [reply Mathlib/Data/Rat/Cast/CharZero.lean:84]: Doesn't feel like a huge golf to me, but okay

**Comment by YaelDillies** [comment]
bors merge

**Comment by mattrobball** [comment]
bors r-

Sorry. The discussion wasn't finished

**Comment by mattrobball** [comment]
@YaelDillies thanks for the offer of help. I appreciate anything you can provide. While I dislike reforming (what I view as) bad design, I understand you have things to move forward. 

bors merge

**Pull request closed without merging**
